### ISSUE TITLE
Use trigger_error instead of error_log

### ISSUE THREAD

**Issue by kkmuffme** [woocommerce/woocommerce#27702]
woocommerce should only be using trigger_error() instead of error_log(), since it's easier to hook into/modify the output location.
Then also the phpcs:ignore can be removed, bc the error is ignored, but the issue is legit, since trigger_error should be used.

**Comment by rrennick** [comment]
@kkmuffme Thanks for the suggestion. The WordPress error / debug logging system is built around `error_log` https://github.com/WordPress/WordPress/blob/master/wp-includes/default-constants.php#L75:L93. While `trigger_error` offers more flexibility it would introduce a lot of additional confusion, administration overhead, etc. to have multiple locations where similar information was logged.

**Comment by lkraav** [comment]
> it would introduce a lot of additional confusion, administration overhead, etc.

Can you clarify with a single example for both cases listed above?

**Comment by kkmuffme** [comment]
@rrennick sorry but you are wrong. all deprecated handling in WP core happens via trigger_error() (see deprecated_argument_run, ...) in /wp-includes/functions.php

But woocommerce in wc-deprecated-functions.php in calling of these functions a) has a separate handling which essentially does what WP core already does thus not sure why woocommerce even has that code and b) uses error_log there.

e.g.

```
function wc_deprecated_function( $function, $version, $replacement = null ) {
	// @codingStandardsIgnoreStart
	if ( is_ajax() || WC()->is_rest_api_request() ) {
		do_action( 'deprecated_function_run', $function, $replacement, $version );
		$log_string  = "The {$function} function is deprecated since version {$version}.";
		$log_string .= $replacement ? " Replace with {$replacement}." : '';
		error_log( $log_string );
	} else {
		_deprecated_function( $function, $version, $replacement );
	}
	// @codingStandardsIgnoreEnd
}
```

WC uses error_log
but the else falls back to WP core function, which itself will call trigger_error.

This means for deprecated code, some will end up in error_log and some in trigger_error.
Which is why I opened this issue to highlight that the usage of error_log is incorrect because it's inconsistent.

**Comment by rrennick** [comment]
> 		error_log( $log_string );

@kkmuffme The reason for using the `error_log()` function in REST API requests is that `_deprecated_function` breaks the response object. WordPress 5.5 added a deprecation to the REST API registration which broke many plugins (including WooCommerce).

**Comment by kkmuffme** [comment]
@rrennick yes, but why use `error_log()` instead of  `trigger_error( $message, E_USER_DEPRECATED );` as the `_deprecated_function` handler does?

**Comment by kkmuffme** [comment]
@rrennick ?

**Comment by rrennick** [comment]
@kkmuffme I've re-opened the issue.

**Comment by jamesckemp** [comment]
This issue has been automatically closed as part of the backlog cleanup announced in [Refining how we manage the WooCommerce backlog](https://developer.woocommerce.com/2026/05/29/backlog-cleanup/). It met the criteria outlined there. If this still affects a current version of WooCommerce, please follow the steps in that post. A fresh report with current reproduction steps is the fastest path to triage.
